ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java

```diff
@@ -0,0 +1,543 @@
+      int gridSizeKernel0Y = (inputDepthImage.cols() + blockSizeXY - 1) / blockSizeXY;
+      int gridSizeKernel1XY = (localCellsPerAxis + blockSizeXY - 1) / blockSizeXY;
+      int gridSizeKernel2XY = (globalCellsPerAxis + blockSizeXY - 1) / blockSizeXY;
+      int gridSizeKernel3XY = (heightMapParameters.getCropWindowSize() + blockSizeXY - 1) / blockSizeXY;
+
+      blockSize = new dim3(blockSizeXY, blockSizeXY, 1);
+      gridSizeKernel0 = new dim3(gridSizeKernel0X, gridSizeKernel0Y, 1);
+      gridSizeKernel1 = new dim3(gridSizeKernel1XY, gridSizeKernel1XY, 1);
+      gridSizeKernel2 = new dim3(gridSizeKernel2XY, gridSizeKernel2XY, 1);
+      gridSizeKernel3 = new dim3(gridSizeKernel3XY, gridSizeKernel3XY, 1);
+
+      preprocessKernel.withPointer(inputDepthImage.data()).withLong(inputDepthImage.step());
+      preprocessKernel.withPointer(transformedInputDepthImage.data()).withLong(transformedInputDepthImage.step());
+
+      updateKernel.withPointer(inputDepthImage.data()).withLong(inputDepthImage.step());
+      updateKernel.withPointer(localHeightMapImage.data()).withLong(localHeightMapImage.step());
+      updateKernel.withPointer(parametersDevicePointer);
+      updateKernel.withPointer(sensorToGroundTransformDevicePointer);
+      updateKernel.withPointer(groundToSensorTransformDevicePointer);
+
+      registerKernel.withPointer(localHeightMapImage.data()).withLong(localHeightMapImage.step());
+      registerKernel.withPointer(globalHeightMapImage.data()).withLong(globalHeightMapImage.step());
+      registerKernel.withPointer(parametersDevicePointer);
+      registerKernel.withPointer(worldToGroundTransformDevicePointer);
+      registerKernel.withPointer(sensorToGroundTransformDevicePointer);
+
+      croppingKernel.withPointer(globalHeightMapImage.data()).withLong(globalHeightMapImage.step());
+      croppingKernel.withPointer(sensorCroppedHeightMapImage.data()).withLong(sensorCroppedHeightMapImage.step());
+      croppingKernel.withPointer(parametersDevicePointer);
+      croppingKernel.withInt(heightMapParameters.getCropWindowSize());
+      cudaStreamSynchronize(stream);
+
+      preprocessKernel.run(stream, gridSizeKernel0, blockSize, 0);
+      updateKernel.run(stream, gridSizeKernel1, blockSize, 0);
+      cudaStreamSynchronize(stream);
+      registerKernel.run(stream, gridSizeKernel2, blockSize, 0);
+
+      cudaStreamSynchronize(stream);
+
+      //Update the terrain map data with the new results
+      terrainMapData.setSensorOrigin(groundToWorldTransform.getTranslationX(), groundToWorldTransform.getTranslationY());
+
+      croppingKernel.run(stream, gridSizeKernel3, blockSize, 0);
+
+      cudaStreamSynchronize(stream);
+
+      Mat finalCroppedHeightMap = new Mat();  // Assuming the height map is 201x201
+      sensorCroppedHeightMapImage.download(finalCroppedHeightMap);  // Download the image from the GPU to the Mat object
+      cudaStreamSynchronize(stream);
```

Comment:
The GpuMat's download function is synchronous so you don't need to synchronize after it (we could use the asynchonous download function, but we haven't figured out how to do that yet)